PR: https://github.com/odoo/odoo/pull/175562

From: ae35752ce7aeb6e62e7ef0ec46c15dd9d23571cc
From: Joris Makauskis
Date: 2024-09-01 04:35:44

Breaking data model changes scores: del:15.399999999999999 + add:1, change matches:
-    account_info_id = fields.Many2one(
-    account_info_ids = fields.One2many(
-    balance = fields.Char(compute='_compute_balance')
-    warn_me = fields.Boolean(
-    warning_email = fields.Char(related='account_info_id.warning_email', readonly=False)
-    show_token = fields.Boolean()
+    balance = fields.Char(readonly=True)
+    warning_user_ids = fields.Many2many('res.users', string="Email Alert Recipients")
-    account_id = fields.Many2one('iap.account', string='IAP Account')
-    account_token = fields.Char()
-    balance = fields.Float(string='Balance', digits=(16, 4), default=0)
-    account_uuid_hashed = fields.Char(string='Account UUID')
-    service_name = fields.Char(string='Related Service')
-    description = fields.Char()
-    warn_me = fields.Boolean('Warn me', default=False)
-    warning_threshold = fields.Float('Threshold')
-    warning_email = fields.Char()

Total Changes: 245

[IMP] iap, sms: improve sms registration

Refactor the iap_account and iap_service models, which allows for a
better registration flow for sms accounts. Where the user can verify
their phone number, set a sender name and warning emails on their own db
without needing to go to the iap website. Additionaly the
assertQueryCount had to be increased for some tests, as the iap.account
is set outside of _contact_iap, which is mocked in those tests.

task-3741280

closes odoo/odoo#175562

Related: odoo/enterprise#67883
Related: odoo/upgrade#6346
Signed-off-by: Louis Baudoux (lba) <lba@odoo.com>

================================= pseudo patch: =================================

--- a/addons/iap/models/__init__.py
+++ b/addons/iap/models/__init__.py
@@ -1,5 +1,5 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import iap_account
 from . import iap_enrich_api
+from . import iap_service

--- a/addons/iap/models/iap_account.py
+++ b/addons/iap/models/iap_account.py
@@ -1,14 +1,15 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import logging
+import secrets
 import threading
 import uuid
 import werkzeug.urls
 
-from odoo import api, fields, models
+from odoo import api, fields, models, _
 from odoo.addons.iap.tools import iap_tools
-from odoo.exceptions import AccessError
+from odoo.exceptions import AccessError, UserError
+from odoo.tools import get_lang
 
 _logger = logging.getLogger(__name__)
 
@@ -17,103 +18,122 @@ DEFAULT_ENDPOINT = 'https://iap.odoo.com'
 
 class IapAccount(models.Model):
     _name = 'iap.account'
-    _rec_name = 'service_name'
     _description = 'IAP Account'
 
     name = fields.Char()
-    service_name = fields.Char(readonly=True)
+    service_id = fields.Many2one('iap.service', required=True)
+    service_name = fields.Char(related='service_id.technical_name')
+    service_locked = fields.Boolean(default=False)  # If True, the service can't be edited anymore
+    description = fields.Char(related='service_id.description')
     account_token = fields.Char(
         default=lambda s: uuid.uuid4().hex,
         help="Account token is your authentication key for this service. Do not share it.",
         size=43)
     company_ids = fields.Many2many('res.company')
-    account_info_id = fields.Many2one(
-        'iap.account.info', compute='_compute_info', inverse='_inverse_info', search='_search_info')
-    account_info_ids = fields.One2many(
-        'iap.account.info', 'account_id',
-        string="Accounts from IAP")
-    balance = fields.Char(compute='_compute_balance')
-    description = fields.Char(related='account_info_id.description')
-    warn_me = fields.Boolean(
-        related='account_info_id.warn_me',
-        help="We will send you an email when your balance gets below that threshold",
-        readonly=False)
-    warning_threshold = fields.Float(related='account_info_id.warning_threshold', readonly=False)
-    warning_email = fields.Char(related='account_info_id.warning_email', readonly=False)
-    show_token = fields.Boolean()
 
-    @api.model
-    def get_view(self, view_id=None, view_type='form', **kwargs):
-        res = super().get_view(view_id, view_type, **kwargs)
-        if view_type == 'tree':
-            self.env['iap.account'].get_services()
-        return res
-
-    @api.depends('account_info_ids')
-    def _compute_info(self):
-        for account in self:
-            if account.account_info_ids:
-                account.account_info_id = account.account_info_ids[-1]
+    # Dynamic fields, which are received from iap server and set when loading the view
+    balance = fields.Char(readonly=True)
+    warning_threshold = fields.Float("Email Alert Threshold")
+    warning_user_ids = fields.Many2many('res.users', string="Email Alert Recipients")
+    state = fields.Selection([('banned', 'Banned'), ('registered', "Registered"), ('unregistered', "Unregistered")], readonly=True)
 
-    @api.depends('account_info_id')
-    def _compute_balance(self):
+    @api.constrains('warning_threshold', 'warning_user_ids')
+    def validate_warning_alerts(self):
         for account in self:
-            account.balance = f'{account.account_info_id.balance} {account.account_info_id.unit_name}' if account.account_info_id else "0 Credits"
+            if account.warning_threshold < 0:
+                raise UserError(_("Please set a positive email alert threshold."))
+            users_with_no_email = [user.name for user in self.warning_user_ids if not user.email]
+            if users_with_no_email:
+                raise UserError(_(
+                    "One of the email alert recipients doesn't have an email address set. Users: %s",
+                    ",".join(users_with_no_email),
+                ))
 
-    def _inverse_info(self):
-        for account in self:
-            if account.account_info_ids:
-                # delete previous reference
-                account_info = account.env['iap.account.info'].browse(account.account_info_ids[0].id)
-                account_info.account_id = False
-            # set new reference
-            account.account_info_id.account_id = account
+    def web_read(self, *args, **kwargs):
+        if not self.env.context.get('disable_iap_fetch'):
+            self._get_account_information_from_iap()
+        return super().web_read(*args, **kwargs)
 
-    def _search_info(self, operator, value):
-        return []
+    def web_save(self, *args, **kwargs):
+        return super(IapAccount, self.with_context(disable_iap_fetch=True)).web_save(*args, **kwargs)
 
     def write(self, values):
-        res = super(IapAccount, self).write(values)
-        iap_edits = ['warn_me', 'warning_threshold', 'warning_email']
-        if any(edited_attribute in values for edited_attribute in iap_edits):
-            try:
-                route = '/iap/update-warning-odoo'
-                endpoint = iap_tools.iap_get_endpoint(self.env)
-                url = endpoint + route
+        res = super().write(values)
+        if (
+            not self.env.context.get('disable_iap_update')
+            and any(warning_attribute in values for warning_attribute in ('warning_threshold', 'warning_user_ids'))
+        ):
+            route = '/iap/1/update-warning-email-alerts'
+            endpoint = iap_tools.iap_get_endpoint(self.env)
+            url = werkzeug.urls.url_join(endpoint, route)
+            for account in self:
                 data = {
-                    'account_token': self.mapped('account_token')[0],
-                    'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
-                    'warn_me': values.get('warn_me'),
-                    'warning_threshold': values.get('warning_threshold'),
-                    'warning_email': values.get('warning_email'),
+                    'account_token': account.account_token,
+                    'warning_threshold': account.warning_threshold,
+                    'warning_emails': [{
+                        'email': user.email,
+                        'lang_code': user.lang or get_lang(self.env).code,
+                    } for user in account.warning_user_ids],
                 }
-                iap_tools.iap_jsonrpc(url=url, params=data)
-            except AccessError as e:
-                _logger.warning('Save service error : %s', str(e))
+                try:
+                    iap_tools.iap_jsonrpc(url=url, params=data)
+                except AccessError as e:
+                    _logger.warning("Update of the warning email configuration has failed: %s", str(e))
         return res
 
-    def get_services(self):
+    @staticmethod
+    def is_running_test_suite():
+        return hasattr(threading.current_thread(), 'testing') and threading.current_thread().testing
+
+    def _get_account_information_from_iap(self):
+        # During testing, we don't want to call the iap server
+        if self.is_running_test_suite():
+            return
+        route = '/iap/1/get-accounts-information'
+        endpoint = iap_tools.iap_get_endpoint(self.env)
+        url = werkzeug.urls.url_join(endpoint, route)
+        params = {
+            'iap_accounts': [{
+                'token': account.account_token,
+                'service': account.service_id.technical_name,
+            } for account in self if account.service_id],
+            'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
+        }
         try:
-            route = '/iap/services-token'
-            endpoint = iap_tools.iap_get_endpoint(self.env)
-            url = endpoint + route
-            account_tokens = self.env['iap.account'].sudo().search([]).mapped('account_token')
-            params = {
-                'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
-                'iap_accounts': account_tokens,
-            }
-            services = iap_tools.iap_jsonrpc(url=url, params=params)
-            for service in services:
-                account_id = self.env['iap.account'].sudo().search(
-                    [('account_token', '=', service['account_token'])]).ids[0]
-                service['account_id'] = account_id
-                self.env['iap.account.info'].create(service)
+            accounts_information = iap_tools.iap_jsonrpc(url=url, params=params)
         except AccessError as e:
-            _logger.warning('Get services error : %s', str(e))
+            _logger.warning("Fetch of the IAP accounts information has failed: %s", str(e))
+            return
+
+        for token, information in accounts_information.items():
+            account_id = self.filtered(lambda acc: secrets.compare_digest(acc.account_token, token))
+
+            # Default rounding of 4 decimal places to avoid large decimals
+            balance_amount = round(information['balance'], None if account_id.service_id.integer_balance else 4)
+            balance = f"{balance_amount} {account_id.service_id.unit_name}"
+
+            information.pop('link_to_service_page', None)
+            account_info = {
+                'balance': balance,
+                'warning_threshold': information['warning_threshold'],
+                'state': information['registered'],
+                'service_locked': True,  # The account exist on IAP, prevent the edition of the service
+            }
+
+            if account_id.service_name == 'sms':
+                account_info.update({
+                    'sender_name': information.get('sender_name')
+                })
+
+            account_id.with_context(disable_iap_update=True, tracking_disable=True).write(account_info)
 
     @api.model_create_multi
     def create(self, vals_list):
         accounts = super().create(vals_list)
+        for account in accounts:
+            if not account.name:
+                account.name = account.service_id.name
+
         if self.env['ir.config_parameter'].sudo().get_param('database.is_neutralized'):
             # Disable new accounts on a neutralized database
             for account in accounts:
@@ -142,9 +162,12 @@ class IapAccount(models.Model):
                 IapAccount.search(domain + [('account_token', '=', False)]).sudo().unlink()
                 accounts = accounts - accounts_without_token
         if not accounts:
-            if hasattr(threading.current_thread(), 'testing') and threading.current_thread().testing:
+            service = self.env['iap.service'].search([('technical_name', '=', service_name)], limit=1)
+            if not service:
+                raise UserError("No service exists with the provided technical name")
+            if self.is_running_test_suite():
                 # During testing, we don't want to commit the creation of a new IAP account to the database
-                return self.create({'service_name': service_name})
+                return self.sudo().create({'service_id': service.id})
 
             with self.pool.cursor() as cr:
                 # Since the account did not exist yet, we will encounter a NoCreditError,
@@ -158,7 +181,7 @@ class IapAccount(models.Model):
                 if not account:
                     if not force_create:
                         return account
-                    account = IapAccount.create({'service_name': service_name})
+                    account = IapAccount.create({'service_id': service.id})
                 # fetch 'account_token' into cache with this cursor,
                 # as self's cursor cannot see this account
                 account_token = account.account_token
@@ -170,6 +193,10 @@ class IapAccount(models.Model):
             return accounts_with_company[0]
         return accounts[0]
 
+    @api.model
+    def get_account_id(self, service_name):
+        return self.get(service_name).id
+
     @api.model
     def get_credits_url(self, service_name, base_url='', credit=0, trial=False, account_token=False):
         """ Called notably by ajax crash manager, buy more widget, partner_autocomplete, sanilmail. """
@@ -177,7 +204,7 @@ class IapAccount(models.Model):
         if not base_url:
             endpoint = iap_tools.iap_get_endpoint(self.env)
             route = '/iap/1/credit'
-            base_url = endpoint + route
+            base_url = werkzeug.urls.url_join(endpoint, route)
         if not account_token:
             account_token = self.get(service_name).account_token
         d = {
@@ -191,18 +218,13 @@ class IapAccount(models.Model):
         return '%s?%s' % (base_url, werkzeug.urls.url_encode(d))
 
     def action_buy_credits(self):
-        for account in self:
-            return {
-                'type': 'ir.actions.act_url',
-                'url': self.env['iap.account'].get_credits_url(
-                    account_token=account.account_token,
-                    service_name=account.service_name,
-                ),
-            }
-
-    def action_toggle_show_token(self):
-        for account in self:
-            account.show_token = not account.show_token
+        return {
+            'type': 'ir.actions.act_url',
+            'url': self.env['iap.account'].get_credits_url(
+                account_token=self.account_token,
+                service_name=self.service_name,
+            ),
+        }
 
     @api.model
     def get_config_account_url(self):
@@ -215,7 +237,7 @@ class IapAccount(models.Model):
             url = f"/odoo/action-iap.iap_account_action/{account.id}?menu_id={menu.id}"
         else:
             url = f"/odoo/action-iap.iap_account_action?menu_id={menu.id}"
-        return  url
+        return url
 
     @api.model
     def get_credits(self, service_name):
@@ -225,7 +247,7 @@ class IapAccount(models.Model):
         if account:
             route = '/iap/1/balance'
             endpoint = iap_tools.iap_get_endpoint(self.env)
-            url = endpoint + route
+            url = werkzeug.urls.url_join(endpoint, route)
             params = {
                 'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
                 'account_token': account.account_token,
@@ -238,20 +260,3 @@ class IapAccount(models.Model):
                 credit = -1
 
         return credit
-
-
-class IAPAccountInfo(models.TransientModel):
-    _name = 'iap.account.info'
-    _description = 'IAP Account Info'
-    _transient_max_hours = 1
-
-    account_id = fields.Many2one('iap.account', string='IAP Account')
-    account_token = fields.Char()
-    balance = fields.Float(string='Balance', digits=(16, 4), default=0)
-    account_uuid_hashed = fields.Char(string='Account UUID')
-    service_name = fields.Char(string='Related Service')
-    description = fields.Char()
-    warn_me = fields.Boolean('Warn me', default=False)
-    warning_threshold = fields.Float('Threshold')
-    warning_email = fields.Char()
-    unit_name = fields.Char(default='Credits')

--- a/None
+++ b/addons/iap/models/iap_service.py
@@ -0,0 +1,18 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class IapService(models.Model):
+    _name = 'iap.service'
+    _description = 'IAP Service'
+
+    name = fields.Char(required=True)
+    technical_name = fields.Char(readonly=True, required=True)
+    description = fields.Char(required=True, translate=True)
+    unit_name = fields.Char(required=True, translate=True)
+    integer_balance = fields.Boolean(required=True)
+
+    _sql_constraints = [
+        ('unique_technical_name', 'UNIQUE(technical_name)', 'Only one service can exist with a specific technical_name')
+    ]
